### DISCUSSION TITLE
Clearer Build Instructions for Newbies

### DISCUSSION THREAD

**Discussion by TannerStocks** [InfiniTimeOrg/InfiniTime#1541] [Q&A] [answered]
I would love to join the community and write apps for InfiniTime. Before I spend time developing, I thought I should at least learn how to build the firmware first. But for the life of me, I can't get a build to work. The instructions that seem to be the most straightforward are found in InfiniTime/doc/buildWithDocker.md. I still don't understand everything in these instructions, so I'm having to make several assumptions that I know could be wrong. I'm not new to software development, but I am a newbie when it comes to Docker and cmake. I've used Linux for years and am not scared of using a terminal, but I'm probably still a beginner in my Linux knowledge. I am also trying to build on a Windows machine. If I could get some help getting started, that would be wonderful. 

After a lot of digging and following instructions found in different parts of this repo, I understand that the simplest way to build the firmware is to use the docker image from Docker Hub. The instructions in buildWithDocker.md starts with a list of dependencies, and it's not clear if I need to install those myself if using the Docker Hub image. 

Here is the process I followed:

- I cloned the InfiniTime repo and opened it in VS Code.
- I ran the command given by Docker Hub in the visual studio terminal to pull the image 
`docker pull infinitime/infinitime-build`
- The instructions in buildWithDocker.md then said to navigate to the project root, which I would assume to be the repository root, but it also gave `cd ./work/Pinetime` as an example, which confused me, since I couldn't find a similar path in the repo.
-  I then ran this command to build the firmware
`docker run --rm -it -v ${PWD}:/sources --user $(id -u):$(id -g) infinitime-build`
- This gives an error saying it doesn't recognize the "id"
`id : The term 'id' is not recognized as the name of a cmdlet, function, script file, or operable program.`

In short, I feel like I'm missing some important context to what I'm trying to do, and would love some help getting over the hurdles. Thanks in advance!

**Comment by FintasticMan** [comment, marked as answer]
> which I would assume to be the repository root

That is correct.

> This gives an error saying it doesn't recognize the "id"

This error happens because the instructions kind of expect you to be running in a Linux environment. That part specifically tells Docker to run as the current user rather than as root. I'm not very familiar with Windows, maybe you can get away with just leaving that `--user` bit out.

Building on Windows isn't supported anywhere near as well as on Linux, so it might be better to try that as well.

**Reply by TannerStocks** [reply]
Ah thanks for the help! Alright I changed the command and it got a lot farther in the build. 
`docker run --rm -it -v ${PWD}:/sources infinitime/infinitime-build /opt/build.sh pinetime-app`
It still had some gmake errors that I'll have to look into (or see if they go away if I use Linux).
Where should the final .dfu file pop out at the end?

**Reply by FintasticMan** [reply]
There will be a dfu .zip file in `build/src`.

**Reply by TannerStocks** [reply]
Alright I got it to work on Linux with this command, which pulls and runs the docker image from Docker Hub.
`sudo docker run --rm -it -v ${PWD}:/sources --user $(id -u):$(id -g) infinitime/infinitime-build`
Also you have to remember to checkout the correct InfiniTime branch you want (after cloning the InfiniTime repo) and checkout the submodules with
`git submodule update --init --recursive`

The final dfu .zip file `pinetime-mcuboot-app-dfu-1.11.0.zip` is in `build/src`, but it looks like the official output is actually copied to `build/output`

Thanks for the help!

**Reply by JF002** [reply]
As @FintasticMan wrote, the instructions expect that you're running in a Linux environment. I do not use windows for any development purposes, so I cannot check that the instructions also work on that OS...

If you think the instructions need some additional information based on your experience, feel free to update them and create a pull-request so we can improve the doc together :)

**Comment by TailyFair** [comment]
I just ordered my PineTime since I want to create custom app and tried to setup InfiniTime.

I found that using simulator is by far the easiest approach to get started. https://github.com/InfiniTimeOrg/InfiniSim
I think it should be mentioned somewhere in this repo, since I think it's irreplaceable  tool.
It worked with 0 prior knowledge by installing WSL2 Ubuntu on Windows 11.
Once built, even GUI worked.
